apps/s_client.c

```diff
@@ -3467,4 +3509,25 @@ static int ldap_ExtendedResponse_parse(const char *buf, long rem)
     return ret;
 }
 
+static char *base64encode (const void *buf, size_t len) {
+
+    /* BASE64 encoder - used only for encoding basic proxy authentication credentials
+     * Based on an idea by schulwitz here: 
+     * https://stackoverflow.com/questions/342409/how-do-i-base64-encode-decode-in-c/
+     */
+
+     BIO *b64_bio, *mem_bio;
+     BUF_MEM *mem_bio_mem_ptr;
+     b64_bio = BIO_new(BIO_f_base64());
+     mem_bio = BIO_new(BIO_s_mem());
+     BIO_push(b64_bio, mem_bio);
+     BIO_set_flags(b64_bio, BIO_FLAGS_BASE64_NO_NL);  /* No newlines */ 
+     BIO_write(b64_bio, buf, len);
+     (void)BIO_flush(b64_bio);
+     BIO_get_mem_ptr(mem_bio, &mem_bio_mem_ptr);  /* Get address of mem_bio memory structure */
+     char *base64_encoded = OPENSSL_zalloc ((*mem_bio_mem_ptr).length + 1); /* +1 oversize for terminating NULL */
+     if (base64_encoded) memcpy(base64_encoded, (*mem_bio_mem_ptr).data, (*mem_bio_mem_ptr).length); /* Copy to return buffer */
```

Comment:
does this work?  you should use accessor functions.

Reply:
the comment on the memcpy and cleanup lines should be removed.

Reply:
It does work, but I've reworked the function using Matt's suggestion below which simplifies things and removes the need for any accessor functions.